### ISSUE TITLE
Moonshine sources built through the Moonshine having problems with opening files on MacOS

### ISSUE THREAD

**Issue by piotrzarzycki21** [prominic/Moonshine-IDE#49]
I wanted to use Moonshine to fix some issues on MacOS(#6, #7). I was able to setup, built and run Moonshine sources ("Build & Debug"). Unfortunately in any kind of places where I would like to change some path to folder/files object File() throws an event Event.CANCEL.

**Scenario:**

1. Run Moonshine and go to global "Settings" -> "Default SDK"
2. Click "Change" and try to "Browse" SDK

**Current Results**
Nothing happens. Browse window do not appear.

Following code has been used for open file browser:
```
_file.addEventListener(Event.SELECT, onSelectHandler);
_file.addEventListener(Event.CANCEL, onCancelHandler);
_file.browseForDirectory(title);
```
Once user click on "Browse" after some time `onCancelHandler` is being called.

I see in the MacOS "Console" following errors:
`​default	14:09:48.283117 +0200	adl	warning: <NSRemoteView: 0x112a60b80> determined it was necessary to configure <NSVBOpenPanel: 0x112a36940> to support remote view vibrancy
default	14:09:48.322590 +0200	adl	+[NSXPCSharedListener endpointForReply:withListenerName:]: an error occurred while attempting to obtain endpoint for listener 'com.apple.view-bridge': Connection interrupted
default	14:09:48.322749 +0200	adl	*** Assertion failure in +[NSXPCSharedListener connectionForListenerNamed:fromServiceNamed:], /Library/Caches/com.apple.xbs/Sources/ViewBridge/ViewBridge-283/NSXPCSharedListener.m:415
default	14:09:48.327052 +0200	adl	-[NSVBSavePanel init] caught non-fatal NSInternalInconsistencyException 'NSXPCSharedListener unable to create endpoint for listener named com.apple.view-bridge' with backtrace (
0 CoreFoundation 0x00007fff821992cb __exceptionPreprocess + 171
1 libobjc.A.dylib 0x00007fff96fb148d objc_exception_throw + 48
2 CoreFoundation 0x00007fff8219e042 +[NSException raise:format:arguments:] + 98
3 Foundation 0x00007fff83be6be0 -[NSAssertionHandler handleFailureInMethod:object:file:lineNumber:description:] + 195
4 ViewBridge 0x00007fff959f2ca1 +[NSXPCSharedListener connectionForListenerNamed:fromServiceNamed:] + 1028
5 ViewBridge 0x00007fff95a06128 -[NSRemoteView serviceMarshalConnection] + 332
6 ViewBridge 0x00007fff95a08261 -[NSRemoteView _advanceToConfigPhaseLegacy] + 606
7 ViewBridge 0x00007fff95a09189 -
default	14:09:48.329215 +0200	adl	warning: <NSRemoteView: 0x112bce000> determined it was necessary to configure <NSVBOpenPanel: 0x112bc3000> to support remote view vibrancy
default	14:09:58.370717 +0200	adl	+[NSXPCSharedListener endpointForReply:withListenerName:]: an error occurred while attempting to obtain endpoint for listener 'com.apple.view-bridge': Connection interrupted
default	14:09:58.370847 +0200	adl	*** Assertion failure in +[NSXPCSharedListener connectionForListenerNamed:fromServiceNamed:], /Library/Caches/com.apple.xbs/Sources/ViewBridge/ViewBridge-283/NSXPCSharedListener.m:415
default	14:09:58.373579 +0200	adl	-[NSVBSavePanel init] caught non-fatal NSInternalInconsistencyException 'NSXPCSharedListener unable to create endpoint for listener named com.apple.view-bridge' with backtrace (
0 CoreFoundation 0x00007fff821992cb __exceptionPreprocess + 171
1 libobjc.A.dylib 0x00007fff96fb148d objc_exception_throw + 48
2 CoreFoundation 0x00007fff8219e042 +[NSException raise:format:arguments:] + 98
3 Foundation 0x00007fff83be6be0 -[NSAssertionHandler handleFailureInMethod:object:file:lineNumber:description:] + 195
4 ViewBridge 0x00007fff959f2ca1 +[NSXPCSharedListener connectionForListenerNamed:fromServiceNamed:] + 1028
5 ViewBridge 0x00007fff95a06128 -[NSRemoteView serviceMarshalConnection] + 332
6 ViewBridge 0x00007fff95a08261 -[NSRemoteView _advanceToConfigPhaseLegacy] + 606
7 ViewBridge 0x00007fff95a09189 -`

In order to check whether it is problem in Moonshine I have setup sources in Intellij on Mac - I was able to work without the problem. 

I would start investigation of that issue from creating simple Adobe Air Flex app which uses File through the Moonshine.

**Comment by rat-moonshine** [comment]
Try changing the `<id>` value to anything than the default (`idToBeReplacedByANT`) one in application descriptor file and test.

**Comment by piotrzarzycki21** [comment]
@rat-moonshine unfortunately it didn't help.

**Comment by rat-moonshine** [comment]
I see the exception also thrown in Moonshine console, as well:

```
: 2017-10-25 16:14:40.080 adl[514:6179] an error occurred while attempting to obtain endpoint for listener 'com.apple.view-bridge': Connection interrupted
: 2017-10-25 16:14:40.080 adl[514:6179] *** Assertion failure in +[NSXPCSharedListener connectionForListenerNamed:fromServiceNamed:], /Library/Caches/com.apple.xbs/Sources/ViewBridge/ViewBridge-159/NSXPCSharedListener.m:405
: 2017-10-25 16:14:40.082 adl[514:6179] -[NSVBSavePanel init] caught non-fatal NSInternalInconsistencyException 'NSXPCSharedListener unable to create endpoint for listener named com.apple.view-bridge' with backtrace (
: 	0   CoreFoundation                      0x00007fff9ac44452 __exceptionPreprocess + 178
: 	1   libobjc.A.dylib                     0x00007fff8f205f7e objc_exception_throw + 48
: 	2   CoreFoundation                      0x00007fff9ac4916a +[NSException raise:format:arguments:] + 106
: 	3   Foundation                          0x00007fff8ef918a6 -[NSAssertionHandler handleFailureInMethod:object:file:lineNumber:description:] + 198
: 	4   ViewBridge                          0x00007fff9d14a69a +[NSXPCSharedListener connectionForListenerNamed:fromServiceNamed:] + 168
: 	5   ViewBridge                          0x00007fff9d15a57f -[NSRemoteViewBase serviceMarshalConnection] + 298
: 	6   ViewBridge                          0x00007fff9d15c00b -[NSRemoteViewBase advanceToConfigPhase:awaitingWindowRights:] + 436
: 	7   ViewBridge                          0x00007fff9d15cec4 -[NSRemoteViewBase viewServiceMarshalProxy:withDetailedErrorHandler:] + 225
: 	8   ViewBridge                          0x00007fff9d15d2a5 -[NSRemoteViewBase viewServiceMarshalProxy:withErrorHandler:] + 99
: 	9   ViewBridge                          0x00007fff9d15158c -[NSRemoteViewBase bridge] + 225
: 	10  AppKit                              0x00007fff8c0ed6b7 -[NSVBSavePanel init] + 389
: 	11  AppKit                              0x00007fff8c0ed4c3 +[NSSavePanel newRemotePanel] + 301
: 	12  AppKit                              0x00007fff8c0ed17b +[NSSavePanel _crunchyRawUnbonedPanel] + 95
: 	13  Adobe AIR_64                        0x0000000101eb0fcf Adobe AIR_64 + 7016399
: 	14  Adobe AIR_64                        0x0000000101eb0d73 Adobe AIR_64 + 7015795
: 	15  Foundation                          0x00007fff8ef16dfe __NSThreadPerformPerform + 279
: 	16  CoreFoundation                      0x00007fff9abd97e1 __CFRUNLOOP_IS_CALLING_OUT_TO_A_SOURCE0_PERFORM_FUNCTION__ + 17
: 	17  CoreFoundation                      0x00007fff9abb8f0c __CFRunLoopDoSources0 + 556
: 	18  CoreFoundation                      0x00007fff9abb842f __CFRunLoopRun + 927
: 	19  CoreFoundation                      0x00007fff9abb7e28 CFRunLoopRunSpecific + 296
: 	20  HIToolbox                           0x00007fff9a7d7935 RunCurrentEventLoopInMode + 235
: 	21  HIToolbox                           0x00007fff9a7d7677 ReceiveNextEventCommon + 184
: 	22  HIToolbox                           0x00007fff9a7d75af _BlockUntilNextEventMatchingListInModeWithFilter + 71
: 	23  AppKit                              0x00007fff8be83df6 _DPSNextEvent + 1067
: 	24  AppKit                              0x00007fff8be83226 -[NSApplication _nextEventMatchingEventMask:untilDate:inMode:dequeue:] + 454
: 	25  AppKit                              0x00007fff8be77d80 -[NSApplication run] + 682
: 	26  Adobe AIR_64                        0x00000001018021fb Adobe AIR_64 + 8699
: 	27  adl                                 0x0000000100001d10 _Z11RuntimeMainPKci + 141
: 	28  adl                                 0x00000001000017d4 start + 52
: 	29  ???                                 0x0000000000000004 0x0 + 4
```
Is it possible that it is because the API is running through a sandbox application (?)

**Comment by rat-moonshine** [comment]
Disabling sandbox confirms the problem. Once I disabled and build a version, I able to use `browseForDirectory` API effectively (after running Moonshine source through Moonshine IDE). It seems sandbox application added some layer that blocks Flash's regular file browsing access. 

Re-enabled sandbox again to Git source.

**Comment by piotrzarzycki21** [comment]
@rat-moonshine in the other words we will not able to do to much, rather disabling sandbox. I will close that one.

**Comment by rat-moonshine** [comment]
I'll suggest keep the issue remain open now. We have some tests discussed with @JoelProminic which we need to check. In other word, this is a more critical problem than it may seem (at this moment). People often use/tests File API, file browsing etc. in their AIR projects. If this continue breaking those project when running in sandbox situation, that could well be a big problem.

We may need to discuss on this more seriously in coming days.

**Comment by JoelProminic** [comment]
To clarify, my idea was to use the Access Manager interface (File > Access Manager) to allow access to the files used by the launched application.

So, if the user does this:
1. Build & Run application
2. Run an action that will open ~/Documents/TestFiles/test.txt
3. Application returns an exception

They could add an entry for ~/Documents/TestFiles in Access Manager, to allow access for their app.  For Filechoosers, the user would need to add access to the default directory, as well as directory that has the files the user needs.  We need to confirm that the sandbox actually works this way.

This is not an ideal solution, but it could be a good workaround.  We would need documentation to make it clear that this feature was available, though.

Alternatively, we could release a copy of Moonshine with the sandbox disabled.  I think we mainly enabled the Sandbox to allow this to work with the App Store.   This still isn't an ideal solution - we would need documentation to instruct users to download the unsandboxed version if they start getting errors like this.

**Comment by rat-moonshine** [comment]
Accessing File API in sandbox environment seems not the problem as we thought about first. I tired to run this piece of codes in Moonshine IDE and this ran well:

```
<fx:Script>
		<![CDATA[
			import flash.filesystem.File;
			
			private var fl:File;
			
			private function onButtonClicked():void
			{
				fl = new File();
				//fl = File.desktopDirectory;
				fl.browseForDirectory("Open");
			}
			
		]]>
	</fx:Script>
	
	<s:Button label="Start" click="onButtonClicked()"/>
```
This shifts our focus to the _IFileBriegeImp_'s `browseForDirectory(..)` command now. I need to debug in Moonshine IDE to determine what is going on when the source ran - but unable to do so due this debug problem - https://github.com/prominic/Moonshine-IDE/issues/47. MXML project falls to this problem due to my last test on this; I may need to wait until Josh solves this problem.

**Comment by piotrzarzycki21** [comment]
@rat-moonshine ,

Maybe try differently:
1) Create class MyClass with one method Test.
2) Create instance of that class in main Application file.
3) Put all the code which you would like to debug in Test method
4) In onButtonClicked run Test method 

Debugger will work in that case.

**Comment by rat-moonshine** [comment]
Yes, that could be an workaround. I can give a test by this soon, @piotrzarzycki21 . Thanks!

**Comment by JoelProminic** [comment]
When I tested @rat-moonshine's code, I got:
```
    : 2017-10-30 15:02:59.156 adl[68668:28435219] an error occurred while attempting to connect to listener 'com.apple.view-bridge': Connection interrupted
    : 2017-10-30 15:02:59.156 adl[68668:28435219] *** Assertion failure in +[NSXPCSharedListener connectionForListenerNamed:fromServiceNamed:], /SourceCache/ViewBridge/ViewBridge-105/NSXPCSharedListener.m:394
    : 2017-10-30 15:02:59.157 adl[68668:28435219] NSXPCSharedListener unable to create endpoint for listener named com.apple.view-bridge
    : 2017-10-30 15:02:59.158 adl[68668:28435219] (
    : 	0   CoreFoundation                      0x00007fff8c5aa03c __exceptionPreprocess + 172
    : 	1   libobjc.A.dylib                     0x00007fff8f85d76e objc_exception_throw + 43
    : 	2   CoreFoundation                      0x00007fff8c5a9e1a +[NSException raise:format:arguments:] + 106
    : 	3   Foundation                          0x00007fff8eada99b -[NSAssertionHandler handleFailureInMethod:object:file:lineNumber:description:] + 195
    : 	4   ViewBridge                          0x00007fff88080af8 +[NSXPCSharedListener connectionForListenerNamed:fromServiceNamed:] + 151
    : 	5   ViewBridge                          0x00007fff8805efc1 -[NSRemoteViewBase serviceMarshalConnection] + 286
    : 	6   ViewBridge                          0x00007fff8805fd10 -[NSRemoteViewBase advanceToConfigPhase:awaitingWindowRights:] + 414
    : 	7   ViewBridge                          0x00007fff8806e5db -[NSWindowCentricRemoteView advanceToConfigPhase] + 947
    : 	8   ViewBridge                          0x00007fff88060883 -[NSRemoteViewBase viewServiceMarshalProxy:withErrorHandler:] + 88
    : 	9   ViewBridge                          0x00007fff88056eb8 -[NSRemoteViewBase bridge] + 207
    : 	10  AppKit                              0x00007fff8ab9b61d -[NSVBSavePanel init] + 303
    : 	11  AppKit                              0x00007fff8ab9b471 +[NSSavePanel newRemotePanel] + 301
    : 	12  AppKit                              0x00007fff8ab9b273 +[NSSavePanel _crunchyRawUnbonedPanel] + 74
    : 	13  Adobe AIR_64                        0x00000001016caa7b Adobe AIR_64 + 7121531
    : 	14  Adobe AIR_64                        0x00000001016ca814 Adobe AIR_64 + 7120916
    : 	15  Foundation                          0x00007fff8eaa4dd0 __NSThreadPerformPerform + 293
    : 	16  CoreFoundation                      0x00007fff8c4c6a01 __CFRUNLOOP_IS_CALLING_OUT_TO_A_SOURCE0_PERFORM_FUNCTION__ + 17
    : 	17  CoreFoundation                      0x00007fff8c4b8c5c __CFRunLoopDoSources0 + 476
    : 	18  CoreFoundation                      0x00007fff8c4b81bf __CFRunLoopRun + 927
    : 	19  CoreFoundation                      0x00007fff8c4b7bd8 CFRunLoopRunSpecific + 296
    : 	20  HIToolbox                           0x00007fff8b5d056f RunCurrentEventLoopInMode + 235
    : 	21  HIToolbox                           0x00007fff8b5d01ee ReceiveNextEventCommon + 179
    : 	22  HIToolbox                           0x00007fff8b5d012b _BlockUntilNextEventMatchingListInModeWithFilter + 71
    : 	23  AppKit                              0x00007fff8a9d78ab _DPSNextEvent + 978
    : 	24  AppKit                              0x00007fff8a9d6e58 -[NSApplication nextEventMatchingMask:untilDate:inMode:dequeue:] + 346
    : 	25  AppKit                              0x00007fff8a9ccaf3 -[NSApplication run] + 594
    : 	26  Adobe AIR_64                        0x00000001010021ec Adobe AIR_64 + 8684
    : 	27  adl                                 0x0000000100002109 _Z11RuntimeMainPKci + 142
    : 	28  adl                                 0x0000000100001bec start + 52
    : 	29  ???                                 0x0000000000000004 0x0 + 4
    : )
```

I would suggest specifying a specific default directory for the filechooser so that it is (hopefully) clear which files are being accessed.

I would also suggest testing by opening (and reading from) files without using a filechooser.  If this is a file access/sandbox issue, then that will help isolate the problem further.

**Comment by rat-moonshine** [comment]
Its possible that I was using the Moonshine IDE that I installed some time back by disabling sandbox for this tests. Because with today's tests, I received same error as of Joel.

To ensure that I was not using sandbox disabled version, I downloaded a new Moonshine development today and installed. Therefore when tried the same `browseForDirectory(...)` code, I received the error as above.

To test if adding the path to Access Manager (in Moonshine IDE) can resolve this problem, I added `~/Users/<user>/Desktop` directory to the Access Manager. Run the desktop project's code as this -

```
fl = File.desktopDirectory;
fl.browseForDirectory("Open");
```

But this was still throwing errors. 

Surprisingly, reading a file goes well (with _given_ path) :

```
private function onButtonClicked():void
{
	var file:File = File.desktopDirectory.resolvePath('a.xml');
	var stream:FileStream = new FileStream();
	stream.open(file, FileMode.READ);
	var document:XML = XML(stream.readUTFBytes(stream.bytesAvailable));

	Alert.show(document.toXMLString());
}
```

**Comment by rat-moonshine** [comment]
I tested further, and things becoming interesting.

In my previous `FileMode.READ/WRITE` tests, I was using user's Desktop directory. Since the desktop directory I already had added in Access Manager in my previous steps, I chose to keep the file that I want to read/write in a different location this time - which was not added to my Access Manager. 

Doing that using read/write command thrown me file access error. 

I opened up Access Manager and added the folder containing the file I want to access, re-run the desktop project from Moonshine, and voila - it able to access the file again. 

So its definite that Moonshine IDE's sandbox affecting the securities of running application, too.

**Comment by rat-moonshine** [comment]
Testing further, I tried to run Moonshine source with the help of its ANE this time - from Moonshine IDE. I did some test changes to the source (so we can build and run with a given ANE (extracted)) and packaged it as a sandbox application. 

After installing from new installer I ran Moonshine, and therefore run the Moonshine source from it. I made it sure that the new ran instance of Moonshine chose ANE instead of ActionScript File API to open the file browser dialogue. 

But it thrown similar error. 

Therefore I created a simple ActionScript project, added Moonshine's ANE in it, and ran. I also ensured that it'll trigger the ANE's file browser upon triggering a button. The run also thrown same errors when clicked on the button. 

Next I ran the same project, now its from the Terminal. With 'adl' I ran the project with ANE support, and this time it able to open the file browser dialogue, with ANE. 

So basically when we ran a project with file browsing ANE from Moonshine IDE - it became problematic to use but from outside.

On the other hand, the ANE usage from Moonshine was a good test case, though. I do not have a basic Mac ANE (which may show some alert only) to test if the procedure works anyway if other features (of ANE). If anybody has any sample Mac or Windows ANE knows which is simple but can be use for test, that would be helpful.

**Comment by rat-moonshine** [comment]
From a basic standpoint, I tried to test the Native Extension integration by using a Mac ANE. I found this one - but its a very old managed one anyway - .

Upon running I received this output in Moonshine console, but no full-screen happened. I'm not sure what else we need to have in the project code, but it looks like its communicating:
```
: [MacOSFullScreen] Initalizing ANE...
: [MacOSFullScreen] ANE initialization complete.
: [MacOSFullScreen] Enabling full screen...
: [MacOSFullScreen] Full screen enabled.
: 2017-10-31 17:40:37.204 adl[19421:128728] setPresentationOptions called with NSApplicationPresentationFullScreen when there is no visible fullscreen window; this call will be ignored.
: [MacOSFullScreen] Enabling Yosemite compatibility.
```
Mine is El Capitan, the ANE looks like supported till Yosemite. 

The bundling procedure works by unzipping extension files manually, at this moment.

**Comment by JoelProminic** [comment]
We should retest this to see if it is still a problem with the non-Sandbox build.

**Comment by rat-moonshine** [comment]
I have created an Wiki page with updated instructions to run Moonshine instance through Moonshine-IDE here: https://github.com/prominic/Moonshine-IDE/wiki/Running-Moonshine-instance-through-Moonshine-IDE.

However, testing with latest Mojave, I see the older file-system error is back (when try opening file-dialog in running instance). I think we haven't noticed this when me and @JoelProminic tested the steps in earlier El Capitan and High Sierra. I don't have any immediate idea on how to overcome this.

**Comment by piotrzarzycki21** [comment]
@rat-moonshine I have finally tired instruction on Windows. All steps were good till I reached `Build & Run` which is failing with following error:

```
: Compiling MoonshineDESKTOPevolved
: Command: "D:\Work\flex-sdk-4.16.1_adobe_air_32\bin\mxmlc.bat" -load-config+=obj/MoonshineDESKTOPevolvedConfig.xml -accessible=true -allow-source-path-overlap=true -swf-version 39 -define+=CONFIG::OSX,false  +configname=air -locale=en_US,ja_JP -debug=true -o bin-debug/MoonshineDESKTOP.swf
: 'C:\Program' is not recognized as an internal or external command,
: operable program or batch file.
: Compilation of MoonshineDESKTOPevolved finished.
```

I think it may be a problem with path to JAVA_HOME. It looks like that:
```
@echo off
set JAVA_HOME=C:\Program Files\Java\jdk8u202-b08
set ANT_HOME=C:\ant
set GRADLE_HOME=C:\MoonshineSDKs\Gradle\gradle-5.4.1-bin
set GRAILS_HOME=C:\MoonshineSDKs\Grails\grails-3.3.10
set FLEX_HOME=D:\Work\flex-sdk-4.16.1_adobe_air_32
set AIR_SDK_HOME=D:\Work\flex-sdk-4.16.1_adobe_air_32
set PATH=%JAVA_HOME%\bin;%ANT_HOME%\bin;%GRADLE_HOME%\bin;%GRAILS_HOME%\bin;%FLEX_HOME%\bin;%AIR_SDK_HOME%\bin;%PATH%
"D:\Work\flex-sdk-4.16.1_adobe_air_32\bin\mxmlc.bat" -load-config+=obj/MoonshineDESKTOPevolvedConfig.xml -accessible=true -allow-source-path-overlap=true -swf-version 39 -define+=CONFIG::OSX,false  +configname=air -locale=en_US,ja_JP -debug=true -o bin-debug/MoonshineDESKTOP.swf
```

**Comment by JoelProminic** [comment]
I'm concerned that @piotrzarzycki21 found another occurrance of the JAVA_HOME spaces issue (i.e. #647).  I updated my JAVA_HOME settings in Moonshine to a path in Program Files, but I was not able to reproduce the "C:\Program" error.

Some issues I noticed with the writeup:

This path should be ...MoonshineDESKTOPevolved/lib/AwayBuilderLib.swc.  I updated this:
> 9. In opening file-browse dialog locate the file Moonshine-IDE/ide/MoonshineDESKTOPevolved/src/libs/AwayBuilderLib.swc and select to Open

There is no mention of changing the <id> in MoonshineDESKTOPevolved-app.xml.  I was able to launch the application, though, so maybe this doesn't matter anymore.

**Comment by rat-moonshine** [comment]
> @rat-moonshine I have finally tired instruction on Windows. All steps were good till I reached Build & Run which is failing with following error:

You really wants to try this on macOS, because the instruction/documentation what intend it to be ;)

@JoelProminic Yes, I know we have space problem with paths during NativeProcess runs on Windows, and this strangely behaves between the command execution. A related issue is already open and in investigation on Git as #641 . Also, thank you for the AwayBuilderLib.swc path fix. 

To be more precise, what I meant by 'strangely behaves between the commands' is - as an example, path with space to Git binaries we found works for few options and not for a few.

Anyway, this continue be a separate issue. @piotrzarzycki21 you can try the instruction if works for you on macOS, hopefully you won't have any problem. This works for Bing and Joel also. We shall going to close the issue in Wednesday meeting.

**Comment by rat-moonshine** [comment]
> I think it may be a problem with path to JAVA_HOME. It looks like that:

During #641  investigation I noticed that path surrounding with double-quotes gives trouble on Windows run. I think same you reported [here](https://github.com/prominic/Moonshine-IDE/issues/647#issuecomment-578835165), too. 

I went on testing the space problem during 'Build & Run' against example Flex Desktop Project, with latest Moonshine build:

Java Location | Project Run
------------------------------- | ----------
C:\MoonshineSDKs\Java\openjava 1.8.0 | Yes
C:\Program Files\Java\openjava 1.8.0 | Yes

I think the compilation and run process shouldn't be any different for Moonshine source. I wonder it acts differently in which certain cases?

**Comment by rat-moonshine** [comment]
@piotrzarzycki21 does your error triggers same for any other Flex desktop project while run? If not, then it perhaps more limited to the Moonshine source running. In that case, you may want to send me the complete project setup you have as a zip, so I can give a try also.

**Comment by piotrzarzycki21** [comment]
> @piotrzarzycki21 does your error triggers same for any other Flex desktop project while run?

It is limited to source code of Moonshine only.

**Comment by piotrzarzycki21** [comment]
> In that case, you may want to send me the complete project setup you have as a zip, so I can give a try also.

What do you exactly want to here ? Moonshine IDE file ?

**Comment by rat-moonshine** [comment]
> > @piotrzarzycki21 does your error triggers same for any other Flex desktop project while run?
> 
> It is limited to source code of Moonshine only.

Please send it over, in a zip.

**Comment by piotrzarzycki21** [comment]
I have finally tried this on my Mac. After setting things up I'm getting following error. Anyone know what do I miss here ?

```
: Compiling MoonshineDESKTOPevolved
: Command: "/Users/piotr/Downloads/flex_4.16.1_air_30/bin/mxmlc" -load-config+=obj/MoonshineDESKTOPevolvedConfig.xml -accessible=true -allow-source-path-overlap=true -define+=CONFIG::OSX,true  +configname=air -locale=en_US,ja_JP -debug=true -o bin-debug/MoonshineDESKTOP.swf
: SDK path: /Users/piotr/Downloads/flex_4.16.1_air_30
: Sending to mxmlc: "/Users/piotr/Downloads/flex_4.16.1_air_30/bin/mxmlc" -load-config+=obj/MoonshineDESKTOPevolvedConfig.xml -accessible=true -allow-source-path-overlap=true -define+=CONFIG::OSX,true  +configname=air -locale=en_US,ja_JP -debug=true -o bin-debug/MoonshineDESKTOP.swf
: Loading configuration file /Users/piotr/Downloads/flex_4.16.1_air_30/frameworks/air-config.xml
: Loading configuration file /Users/piotr/Downloads/Prominic.NET/Moonshine-IDE/ide/MoonshineDESKTOPevolved/obj/MoonshineDESKTOPevolvedConfig.xml
: /Users/piotr/Downloads/Prominic.NET/Moonshine-IDE/ide/MoonshineDESKTOPevolved/src/actionScripts/impls/IFileBridgeImp.as(65): col: 21 Error: Nie znaleziono typu lub nie jest on stałą czasu kompilacji: Main.
: 			private var _ssb:Main = new Main();
: 			                 ^
: /Users/piotr/Downloads/Prominic.NET/Moonshine-IDE/ide/MoonshineDESKTOPevolved/src/actionScripts/impls/IFileBridgeImp.as(65): col: 32 Error: Wywołanie potencjalnie niezdefiniowanej metody Main.
: 			private var _ssb:Main = new Main();
: 			                            ^
: /Users/piotr/Downloads/Prominic.NET/Moonshine-IDE/ide/MoonshineDESKTOPevolved/src/actionScripts/impls/IFileBridgeImp.as(65): col: 32 Error: Wywołanie potencjalnie niezdefiniowanej metody Main.
: 			private var _ssb:Main = new Main();
: 			                            ^
: /Users/piotr/Downloads/Prominic.NET/Moonshine-IDE/ide/MoonshineDESKTOPevolved/src/actionScripts/impls/IFileBridgeImp.as(38): col: 44 Error: Nie znaleziono definicji net.prominic.SecurityScopeBookmark:Main.
: 		import net.prominic.SecurityScopeBookmark.Main;
: 		                                         ^
: Compilation of MoonshineDESKTOPevolved finished.
```

**Comment by rat-moonshine** [comment]
It says you're still compiling stuff as 'sandbox' while trying to use bookmarking ANE. Please, make sure you follow the steps detailed in "Prerequisite before running Moonshine through Moonshine-IDE" at [here](https://github.com/prominic/Moonshine-IDE/wiki/Running-Moonshine-instance-through-Moonshine-IDE).

Additionally, make sure you already fulfilled this need,

> This requires running of non-Sandbox (non-App-Store) version of Moonshine-IDE instance on macOS. You can download non-Sandbox Moonshine-IDE from here: https://moonshine-ide.com/download.

**Comment by piotrzarzycki21** [comment]
@rat-moonshine I have missed one step, sorry about that. Everything is working to me. I was able to build and run Moonshine etc. I'm closing this. Let's not loose anymore time on that.
